chips/cc26x2/src/udma.rs

```diff
@@ -0,0 +1,147 @@
+use kernel::common::registers::{ReadOnly, ReadWrite, WriteOnly};
+use kernel::common::StaticRef;
+use memory_map::UDMA0_BASE;
+
+pub struct UDMARegisters {
+    _status: ReadOnly<u32, Stat::Register>,
+    _cfg: WriteOnly<u32, Cfg::Register>,
+    _ctrl: ReadWrite<u32, Ctrl::Register>,
+    _alt_ctrl: ReadOnly<u32>,
+    _wait_on_req: ReadOnly<u32>,
+    _soft_req: ReadOnly<u32>,
+    set_burst: ReadWrite<u32, Channels::Register>,
+    clear_burst: WriteOnly<u32, Channels::Register>,
+    set_req_mask: ReadWrite<u32, Channels::Register>,
+    clear_req_mask: WriteOnly<u32, Channels::Register>,
+    _set_channel_len: ReadWrite<u32, Channels::Register>,
+    _clear_channel_len: WriteOnly<u32, Channels::Register>,
+    set_channel_pri_alt: ReadWrite<u32, Channels::Register>,
+    clear_channel_pri_alt: WriteOnly<u32, Channels::Register>,
+    set_channel_pri: ReadWrite<u32, Channels::Register>,
+    clear_channel_pri: WriteOnly<u32, Channels::Register>,
+    _reserved: [ReadOnly<u32>; 3],
+    _error: ReadWrite<u32, Error::Register>,
+    _reserved2: [u8; 0x4b4],
+    _req_done: ReadWrite<u32>,
+    _reserved3: [u8; 0x18],
+    _done_mask: ReadWrite<u32>,
+}
+
+register_bitfields! [
+    u32,
+    Stat [
+        TEST            OFFSET(28) NUMBITS(4) [],
+        // Reserved 27:21
+        TOTAL_CHANNELS  OFFSET(21) NUMBITS(6) [],
+        // Reserved 15:8
+        STATE           OFFSET(4) NUMBITS(4) [],
+        // Reserved 3:1
+        MASTER_ENABLE   OFFSET(0) NUMBITS(1) []
+    ],
+    Cfg [
+        // Reserved 31:8
+        PROTO_CTRL      OFFSET(5) NUMBITS(3) [],
+        // Reserved 4:1
+        MASTER_Enable   OFFSET(0) NUMBITS(1) []
+    ],
+    Ctrl [
+        BASE_PTR        OFFSET(10) NUMBITS(21) []
+        // Reserved 9:0
+    ],
+    Channels [
+        TOGGLE          OFFSET(0) NUMBITS(32) []
+    ],
+    Error [
+        STATUS          OFFSET(0) NUMBITS(1) []
+    ]
+];
+
+const UDMA_REG: StaticRef<UDMARegisters> =
+    unsafe { StaticRef::new(UDMA0_BASE as *const UDMARegisters) };
+
+pub const UDMA: Udma = Udma::new();
+
+pub enum Attribute {
+    UseBurst,
+    AltSelect,
+    HighPriority,
+    ReqMask,
+}
+
+pub struct Udma {
+    regs: StaticRef<UDMARegisters>,
+}
+
+impl Udma {
+    pub const fn new() -> Udma {
+        Udma { regs: UDMA_REG }
+    }
+
+    pub fn attribute_enable(&self, channel_num: u32, attribute: Attribute) {
+        let channel_en = 1 << channel_num;
```

Comment:
I would explicitly define the channels in the TOGGLE bitfield so you can use type-safety instead of helper-functions:
```
Channel [
        _0          OFFSET(0) NUMBITS(1) [
            ENABLE = 0x1
            DISABLE = 0x0;
        ]
...
        _19          OFFSET(19) NUMBITS(1) [
            ENABLE = 0x1
            DISABLE = 0x0;
        ]
]
```

This could make all your functions below disappear, as you could instead do:

`self.regs.set_burst.write(CHANNEL._0.ENABLE);`